Source/ProgressionSystemRuntime/Private/Data/PSSaveGameData.cpp

```diff
@@ -53,30 +56,54 @@ void UPSSaveGameData::UnlockLevelByName(FName RowName)
+	FPSSaveToDiskData* CurrentSaveToDiskDataRowPtr = ProgressionSettingsRowDataInternal.Find(CurrentRowName);
+	if (!CurrentSaveToDiskDataRowPtr)
 	{
-		// Increase the current level's progression by the reward from the end game state
-		FName CurrentRowName = UPSWorldSubsystem::Get().GetCurrentRowName();
-		FPSSaveToDiskData* CurrentSaveToDiskDataRowRef = ProgressionSettingsRowDataInternal.Find(CurrentRowName);
-		const FPSRowData& CurrentProgressionSettingsRowData = UPSWorldSubsystem::Get().GetCurrentProgressionSettingsRowByName();
+		return;
+	}
 
-		// do nothing if max start achieved. Max stars of level is amount of point to unlock for a level
-		if (CurrentSaveToDiskDataRowRef->CurrentLevelProgression >= CurrentProgressionSettingsRowData.PointsToUnlock)
-		{
-			return;
-		}
+	// Increase the current level's progression by the reward from the end game state
+	const FPSRowData& CurrentProgressionSettingsRowData = UPSWorldSubsystem::Get().GetCurrentProgressionSettingsRowByName();
+	if (!ensureMsgf(CurrentProgressionSettingsRowData.Character.IsValid(), TEXT("ASSERT: [%i] %hs:\n'CurrentProgressionSettingsRowData or Points to unlock = 0' is not valid!"), __LINE__, __FUNCTION__))
+	{
+		return;
+	}
 
-		const float NewProgression = CurrentSaveToDiskDataRowRef->CurrentLevelProgression + GetProgressionReward(EndGameState);
-		CurrentSaveToDiskDataRowRef->CurrentLevelProgression = FMath::Min(NewProgression, CurrentProgressionSettingsRowData.PointsToUnlock);
+	// do nothing if max start achieved. Max stars of level is amount of point to unlock for a level
+	if (CurrentSaveToDiskDataRowPtr->CurrentLevelProgression >= CurrentProgressionSettingsRowData.PointsToUnlock)
+	{
+		return;
+	}
 
-		// Check if the current level progression has reached or surpassed the points needed to unlock
-		if (CurrentSaveToDiskDataRowRef->CurrentLevelProgression >= CurrentProgressionSettingsRowData.PointsToUnlock)
-		{
-			NextLevelProgressionRowData(); // Advance to the next level's progression data
-		}
-		UPSWorldSubsystem::Get().SaveDataAsync(); // Asynchronously save the updated data
+	const float ProgressionReward = GetProgressionReward(EndGameState);
+
+	// skins to unlock
+	const int32 DataAssetInterval = UPSDataAsset::Get().GetSkinUnlockInterval();
+	if (!ensureMsgf(DataAssetInterval > 0.f, TEXT("ASSERT: [%i] %hs:\n'DataAssetInterval = 0' is not valid!"), __LINE__, __FUNCTION__))
+	{
+		return;
+	}
+	const int32 OldAmountOfUnlockedSkins = FMath::FloorToInt(CurrentSaveToDiskDataRowPtr->CurrentLevelProgression);
+	const float LeftoverStars = FMath::Fmod(CurrentSaveToDiskDataRowPtr->CurrentLevelProgression, DataAssetInterval);
+	const float TotalStars = LeftoverStars + ProgressionReward;
+	const int32 NewSkins = FMath::FloorToInt(TotalStars / DataAssetInterval);
+	const int32 TotalUnlockedSkins = OldAmountOfUnlockedSkins + NewSkins;
+	const int32 MaxUnlockedSkins = UPSWorldSubsystem::Get().GetCurrentSpot()->GetMeshChecked().GetSkinTexturesNum();
```

Comment:
Current Spot might crash here if save is called faster than spot is set and in general we must always handle pointers in c++: please add ensuremsgf on beginning

Reply:
@moinm3uw

Reply:
added : https://github.com/moinm3uw/ProgressionSystem/pull/59/commits/68a5e244deb545b0955c8e66d2154d2845827444